### ISSUE TITLE
dart2js emits code with an error when tearing off an extension method

### ISSUE THREAD

**Issue by alextekartik** [dart-lang/sdk#42712]
This can be observed on the following travis job which run fine on stable (and used to work fine on beta/dev): https://travis-ci.org/github/tekartik/sembast.dart/builds/708424438

It can also be reproduced using this command in this folder: https://github.com/tekartik/sembast.dart/tree/master/sembast:

```
$ pub run test -p chrome test/store_api_test.dart
Compiled 9,379,109 characters Dart to 1,117,802 characters JavaScript in 3.90 seconds
Dart file /tmp/dart_test_VPQTUS/runInBrowser.dart compiled to JavaScript: /tmp/dart_test_REOQQW/test_ZJMTPM/store_api_test.dart.browser_test.dart.js
(hung)
```

However the following works fine:
```
$ pub run build_runner test -- -p chrome test/store_api_test.dart
$ pub run test test/store_api_test.dart
```

This seems to happen on beta only since 2.9.0-21.4.beta and only for this given test file and even if I add `skip: true`

After some investigation, this is the simplest test showing the issue:

```dart
import 'package:sembast/sembast.dart';
import 'package:test/test.dart';

void main() {
  test('hang', () {
    // ignore: unnecessary_statements
    StoreRef.main().record(1).get;
  });
}
```

I agree it is a weird test since get is a function and here it is not calling it however this works fine on stable (2.8.4).

Even this is causing to hang (although marked as skipped) so maybe it is a dart2js issue:

```dart
import 'package:sembast/sembast.dart';
import 'package:test/test.dart';

void main() {
  test('hang', () {
    // ignore: unnecessary_statements
    StoreRef.main().record(1).get;
  }, skip: true);
}
```

**Comment by natebosch** [comment]
Example input:

```dart
extension E<T> on List<T> {
  Future<T> something() async => first;
}

void main() {
  print('Start');
  [1].something;
  print('Done');
}
```

The compiled JS has:

```js
  // ---- 8< -------
  F.E_get_something_closure.prototype = {
    call$0: function() {
      return this.$call$body$E|get#something_closure(this.T._eval$1("0*"));
    },
    $call$body$E|get#something_closure: function($async$type) {
      var $async$goto = 0,
        $async$completer = P._makeAsyncAwaitCompleter($async$type),
        $async$returnValue, $async$self = this;
      var $async$call$0 = P._wrapJsFunctionForAsync(function($async$errorCode, $async$result) {
        if ($async$errorCode === 1)
          return P._asyncRethrow($async$result, $async$completer);
        while (true)
          switch ($async$goto) {
            case 0:
              // Function start
              $async$returnValue = F.E_something($async$self._this, $async$self.T._eval$1("0*"));
              // goto return
              $async$goto = 1;
              break;
            case 1:
              // return
              return P._asyncReturn($async$returnValue, $async$completer);
          }
      });
      return P._asyncStartSync($async$call$0, $async$completer);
    },
    $signature: function() {
      return this.T._eval$1("Future<0*>*()");
    }
  };
  // ---- 8< -------
```

On the 3rd line of this snippet chrome reports an error:

```
Uncaught SyntaxError: Unexpected identifier
```

**Comment by natebosch** [comment]
cc @sigmundch @rakudrama

**Comment by sigmundch** [comment]
This appears similar to #42531, which @rakudrama fixed last week in a38e265f04

I was able to run the little sample at master and it seemed like the program is running properly after his fix. @rakudrama any reason why this would have been working in 2.8?

**Comment by rakudrama** [comment]
Nate's program runs fine at tip-of-tree.

Why did it work with 2.8? With NNBD, we now pass in `T*`, which is no longer simply copying the type, so we need an entry method to compute `T*`, and $body method. In 2.8 we passed in `T` directly, so entry and the body were fused into one method, eliminating the intermediate $body method with a broken name. It was broken in 2.8 for a non-pass-through type, e.g. `Set<T>`:

```dart
extension E<T> on List<T> {
  Future<Set<T>> something() async => {first};
}
```

**Comment by sigmundch** [comment]
Thanks, closing since this was fixed by a38e265

@alextekartik this is not yet available on the beta/dev releases, but we have filed a request to cherry-pick the fix (#42735)

**Comment by alextekartik** [comment]
Thanks @sigmundch 

Cossing fingers as it basically breaks the package for the web and flutter web (which has to be in the beta channel)

**Comment by alextekartik** [comment]
I can confirm the issue got fixed using the latest beta (2.9.0-21.10.beta) release (not dev channel yet though). Thanks again!
